pil-analyzer/src/condenser.rs

```diff
@@ -392,8 +439,43 @@ impl<'a, T: FieldElement> SymbolLookup<'a, T> for Condenser<'a, T> {
                 .new_constraints
                 .push(to_constraint(&constraints, source)),
         }
+        if !self.new_constraints.is_empty() {
+            self.set_non_fresh();
+        }
         Ok(())
     }
+
+    fn capture_stage(&mut self, fun: Arc<Value<'a, T>>) -> Result<Arc<Value<'a, T>>, EvalError> {
+        if !self.stage_is_fresh {
+            return Err(EvalError::InvalidState("Cannot call \"capture_stage\": There are pre-existing constraints or witness columns.".to_string()));
+        }
+
+        let degree = evaluate_function_call(fun, vec![], self)?;
+        let Value::Integer(degree) = degree.as_ref() else {
+            panic!("Type error")
+        };
+        let degree: u64 = degree.try_into().unwrap();
+        if self.degree.is_some() && self.degree != Some(degree) {
+            return Err(EvalError::InvalidState(format!(
+                    "Call to \"capture_stage\" returned degree {degree}, but the degree has already been set to {}.",
+                    self.degree.unwrap()
+                )));
+        } else {
+            self.degree = Some(degree);
+        }
+        self.stage += 1;
+        self.stage_is_fresh = true;
+
+        // TODO use degree
```

Comment:
here or later?

Reply:
Ah this is a leftover, the degree is processed now right above.
```suggestion
```